go-controller/pkg/kubevirt/dhcp.go

```diff
@@ -0,0 +1,138 @@
+}
+
+func composeDHCPConfigs(k8scli *factory.WatchFactory, controllerName string, vmKey ktypes.NamespacedName, podIPs []*net.IPNet) (*dhcpConfigs, error) {
+	if len(podIPs) == 0 {
+		return nil, fmt.Errorf("missing podIPs to compose dhcp options")
+	}
+	if vmKey.Name == "" {
+		return nil, fmt.Errorf("missing vmName to compose dhcp options")
+	}
+
+	dnsServerIPv4, dnsServerIPv6, err := retrieveDNSServiceClusterIPs(k8scli)
+	if err != nil {
+		return nil, fmt.Errorf("failed retrieving dns service cluster ip: %v", err)
+	}
+
+	dhcpConfigs := &dhcpConfigs{}
+	for _, ip := range podIPs {
+		_, cidr, err := net.ParseCIDR(ip.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed converting podIPs to cidr to configure dhcp: %v", err)
+		}
+		if utilnet.IsIPv4CIDR(cidr) {
+			dhcpConfigs.V4 = ComposeDHCPv4Options(cidr.String(), dnsServerIPv4, controllerName, vmKey)
+		} else if utilnet.IsIPv6CIDR(cidr) {
+			dhcpConfigs.V6 = ComposeDHCPv6Options(cidr.String(), dnsServerIPv6, controllerName, vmKey)
+		}
+	}
+	return dhcpConfigs, nil
+}
+
+func retrieveDNSServiceClusterIPs(k8scli *factory.WatchFactory) (string, string, error) {
+	dnsServer, err := k8scli.GetService(config.Kubernetes.DNSServiceNamespace, config.Kubernetes.DNSServiceName)
+	if err != nil {
+		return "", "", err
+	}
+	clusterIPv4 := ""
+	clusterIPv6 := ""
+	for _, clusterIP := range dnsServer.Spec.ClusterIPs {
+		if utilnet.IsIPv4String(clusterIP) {
+			clusterIPv4 = clusterIP
+		} else if utilnet.IsIPv6String(clusterIP) {
+			clusterIPv6 = clusterIP
+		}
+	}
+	return clusterIPv4, clusterIPv6, nil
+}
+
+func ComposeDHCPv4Options(cidr, dnsServer, controllerName string, vmKey ktypes.NamespacedName) *nbdb.DHCPOptions {
+	serverMAC := util.IPAddrToHWAddr(net.ParseIP(ARPProxyIPv4)).String()
+	dhcpOptions := &nbdb.DHCPOptions{
+		Cidr: cidr,
+		Options: map[string]string{
+			"lease_time": fmt.Sprintf("%d", dhcpLeaseTime),
+			"router":     ARPProxyIPv4,
+			"dns_server": dnsServer,
+			"server_id":  ARPProxyIPv4,
+			"server_mac": serverMAC,
+			"hostname":   fmt.Sprintf("%q", vmKey.Name),
+		},
+	}
+	return composeDHCPOptions(controllerName, vmKey, dhcpOptions)
+}
+
+func ComposeDHCPv6Options(cidr, dnsServer, controllerName string, vmKey ktypes.NamespacedName) *nbdb.DHCPOptions {
+	serverMAC := util.IPAddrToHWAddr(net.ParseIP(ARPProxyIPv6)).String()
+	dhcpOptions := &nbdb.DHCPOptions{
+		Cidr: cidr,
+		Options: map[string]string{
+			"server_id": serverMAC,
+		},
+	}
+	if dnsServer != "" {
+		dhcpOptions.Options["dns_server"] = dnsServer
+	}
+	return composeDHCPOptions(controllerName, vmKey, dhcpOptions)
+}
+
+func composeDHCPOptions(controllerName string, vmKey ktypes.NamespacedName, dhcpOptions *nbdb.DHCPOptions) *nbdb.DHCPOptions {
+	dhcpvOptionsDbObjectID := libovsdbops.NewDbObjectIDs(libovsdbops.VirtualMachineDHCPOptions, controllerName,
+		map[libovsdbops.ExternalIDKey]string{
+			libovsdbops.ObjectNameKey: vmKey.String(),
+			libovsdbops.CIDRKey:       dhcpOptions.Cidr,
```

Comment:
@jcaamano one question: I noticed the separator for the keys is a colon - `:` . Will this work for an IPv6 address ?

Reply:
I think this works **today** because we are not doing any parsing but:

- it will not be nice in the eyes of the viewer
- we won't be able to parse this correctly in the future if the need arises

So I thing we should be be replacing `:` with something else in the key value.

Reply:
> I think this works **today** because we are not doing any parsing but:
> 
> * it will not be nice in the eyes of the viewer
> * we won't be able to parse this correctly in the future if the need arises
> 
> So I think we should be be replacing `:` with something else in the key value.

Is difficult to decide what character is not going to collide with the content, I would go for the gold and serialize a slice of strings into a proper format, something that is fast on serialization.

Reply:
wait, one thing I never understood if why we actually want to use cidr as a key. Can you point me to that discussion ? 

Wouldn't we achieve the same end-goal if we used the network name as an index ?

Reply:
At default Network you have multiple cidrs per node.

Reply:
Of course. Thanks.

Reply:
would node name then make sense ? (just trying to come up w/ options).

Reply:
The CIDR is part of the DHCPOptions but node name is not also for future multi-homing use the CIDR will be share across nodes.

Reply:
I am going to replace IPv6 ":" with "_", so just spliting by ":" will give us the correct of elements on the key tuple.